<!DOCTYPE html>
<html lang="en">
  

  <head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width,minimum-scale=1">

  <title>Tabula muris</title>
  <meta name="description" content="Tabula Muris">

  <link rel="canonical" href="http://localhost:4000/scRNA-python-workshop/preprocessing/00-tabula-muris.html">
  <link rel="alternate" type="application/rss+xml" title="Analysis of single cell RNA-seq data" href="http://localhost:4000/scRNA-python-workshop/feed.xml">

  <meta property="og:url"         content="http://localhost:4000/scRNA-python-workshop/preprocessing/00-tabula-muris.html" />
<meta property="og:type"        content="article" />
<meta property="og:title"       content="Tabula muris" />
<meta property="og:description" content="Tabula Muris" />
<meta property="og:image"       content="" />


  <script type="application/ld+json">
  {
  "@context": "http://schema.org",
  "@type": "NewsArticle",
  "mainEntityOfPage":
    "http://localhost:4000/scRNA-python-workshop/preprocessing/00-tabula-muris.html",
  "headline":
    "Tabula muris",
  "datePublished":
    "2019-09-23T08:46:14-07:00",
  "dateModified":
    "2019-09-23T08:46:14-07:00",
  "description":
    "Tabula Muris",
  "author": {
    "@type": "Person",
    "name": "Vladimir Kiselev & the Hemberg lab; adapted by Sidney Bell"
  },
  "publisher": {
    "@type": "Organization",
    "name": "Data 100 at UC Berkeley",
    "logo": {
      "@type": "ImageObject",
      "url": "http://localhost:4000/scRNA-python-workshop",
      "width": 60,
      "height": 60
    }
  },
  "image": {
    "@type": "ImageObject",
    "url": "http://localhost:4000/scRNA-python-workshop",
    "height": 60,
    "width": 60
  }
}

  </script>
  <link rel="stylesheet" href="/scRNA-python-workshop/assets/css/styles.css">
  <link rel="stylesheet" href="https://stackpath.bootstrapcdn.com/font-awesome/4.7.0/css/font-awesome.min.css ">
  <link rel="apple-touch-icon" sizes="57x57" href="/apple-touch-icon-57x57.png">
  <link rel="apple-touch-icon" sizes="60x60" href="/apple-touch-icon-60x60.png">
  <link rel="apple-touch-icon" sizes="72x72" href="/apple-touch-icon-72x72.png">
  <link rel="apple-touch-icon" sizes="76x76" href="/apple-touch-icon-76x76.png">
  <link rel="apple-touch-icon" sizes="114x114" href="/apple-touch-icon-114x114.png">
  <link rel="apple-touch-icon" sizes="120x120" href="/apple-touch-icon-120x120.png">
  <link rel="apple-touch-icon" sizes="144x144" href="/apple-touch-icon-144x144.png">
  <link rel="apple-touch-icon" sizes="152x152" href="/apple-touch-icon-152x152.png">
  <link rel="apple-touch-icon" sizes="180x180" href="/apple-touch-icon-180x180.png">

  <!-- <link rel="manifest" href="/manifest.json"> -->
  <!-- <link rel="mask-icon" href="/safari-pinned-tab.svg" color="#efae0a"> -->
  <meta name="msapplication-TileColor" content="#da532c">
  <meta name="msapplication-TileImage" content="/mstile-144x144.png">
  <meta name="theme-color" content="#233947">

  <!-- Favicon -->
  <link rel="shortcut icon" type="image/x-icon" href="/scRNA-python-workshop/images/logo/favicon.ico">

  <!-- MathJax Config -->
  <!-- Allow inline math using $ and automatically break long math lines -->
<script type="text/x-mathjax-config">
MathJax.Hub.Config({
    tex2jax: {
        inlineMath: [ ['$','$'], ["\\(","\\)"] ],
        displayMath: [ ['$$','$$'], ["\\[","\\]"] ],
        processEscapes: true,
        processEnvironments: true
    },
    CommonHTML: {
        linebreaks: {
            automatic: true,
        },
    },
});
</script>
<script src='https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js?config=TeX-AMS_CHTML' async></script>

  <!-- DOM updating function -->
  <script>
const runWhenDOMLoaded = cb => {
  if (document.readyState != 'loading') {
    cb()
  } else if (document.addEventListener) {
    document.addEventListener('DOMContentLoaded', cb)
  } else {
    document.attachEvent('onreadystatechange', function() {
      if (document.readyState == 'complete') cb()
    })
  }
}

// Helper function to init things quickly
initFunction = function(myfunc) {
  runWhenDOMLoaded(myfunc);
  document.addEventListener('turbolinks:load', myfunc);
};
</script>

  <!-- Define some javascript variables that will be useful in other javascript -->
  <script>
    const site_basename = '/scRNA-python-workshop';
  </script>

  <!-- Add AnchorJS to let headers be linked -->
  <script src="/scRNA-python-workshop/assets/js/anchor.min.js"  type="text/javascript"></script>
  <script>

initFunction(function () {
    anchors.add("main h1, main h2, main h3, main h4")
});

</script>

  <!-- Include Turbolinks to make page loads fast -->
  <!-- https://github.com/turbolinks/turbolinks -->
  <script src="/scRNA-python-workshop/assets/js/turbolinks.js" async></script>
  <meta name="turbolinks-cache-control" content="no-cache">

  <!-- Load nbinteract for widgets -->
  

  <!-- Load Thebelab for interactive widgets -->
  <!-- Include Thebelab for interactive code if it's enabled -->



  <!-- Load the auto-generating TOC -->
  <script src="/scRNA-python-workshop/assets/js/tocbot.min.js"  type="text/javascript"></script>
  <script>
var initToc = function () {
  tocbot.init({
    tocSelector: 'nav.onthispage',
    contentSelector: '.c-textbook__content',
    headingSelector: 'h2, h3',
    orderedList: false,
    collapseDepth: 6,
    listClass: 'toc__menu',
    activeListItemClass: "",  // Not using
    activeLinkClass: "", // Not using
  });
  tocbot.refresh();
}
initFunction(initToc);
</script>

  <!-- Google analytics -->
  <script src="/scRNA-python-workshop/assets/js/ga.js" async></script>

  <!-- Clipboard copy button -->
  <script src="https://cdn.jsdelivr.net/npm/clipboard@2/dist/clipboard.min.js" async></script>

  <!-- Load JS that depends on site variables -->
  <script>
/**
 * Set up copy/paste for code blocks
 */
const codeCellId = index => `codecell${index}`

const clipboardButton = id =>
  `<a id="copy-button-${id}" class="btn copybtn o-tooltip--left" data-tooltip="Copy" data-clipboard-target="#${id}">
    <img src="/scRNA-python-workshop/assets/images/copy-button.svg" alt="Copy to clipboard">
  </a>`

// Clears selected text since ClipboardJS will select the text when copying
const clearSelection = () => {
  if (window.getSelection) {
    window.getSelection().removeAllRanges()
  } else if (document.selection) {
    document.selection.empty()
  }
}

// Changes tooltip text for two seconds, then changes it back
const temporarilyChangeTooltip = (el, newText) => {
  const oldText = el.getAttribute('data-tooltip')
  el.setAttribute('data-tooltip', newText)
  setTimeout(() => el.setAttribute('data-tooltip', oldText), 2000)
}

const addCopyButtonToCodeCells = () => {
  // If ClipboardJS hasn't loaded, wait a bit and try again. This
  // happens because we load ClipboardJS asynchronously.
  if (window.ClipboardJS === undefined) {
    setTimeout(addCopyButtonToCodeCells, 250)
    return
  }

  const codeCells = document.querySelectorAll('div.c-textbook__content > div.highlighter-rouge > div.highlight > pre, div.input_area pre')
  codeCells.forEach((codeCell, index) => {
    const id = codeCellId(index)
    codeCell.setAttribute('id', id)
    if (document.getElementById("copy-button" + id) == null) {
      codeCell.insertAdjacentHTML('afterend', clipboardButton(id));
    }
  })

  const clipboard = new ClipboardJS('.copybtn')
  clipboard.on('success', event => {
    clearSelection()
    temporarilyChangeTooltip(event.trigger, 'Copied!')
  })

  clipboard.on('error', event => {
    temporarilyChangeTooltip(event.trigger, 'Failed to copy')
  })

  // Get rid of clipboard before the next page visit to avoid memory leak
  document.addEventListener('turbolinks:before-visit', () =>
    clipboard.destroy()
  )
}

initFunction(addCopyButtonToCodeCells);
</script>


  <!-- Hide cell code -->
  <script>
    /**
    Add buttons to hide code cells
    */


    var setCodeCellVisibility = function (inputField, kind) {
        // Update the image and class for hidden
        var id = inputField.getAttribute('data-id');
        var codeCell = document.querySelector(`#${id} div.highlight`);

        if (kind === "visible") {
            codeCell.classList.remove('hidden');
            inputField.checked = true;
        } else {
            codeCell.classList.add('hidden');
            inputField.checked = false;
        }
    }

    var toggleCodeCellVisibility = function (event) {
        // The label is clicked, and now we decide what to do based on the input field's clicked status
        if (event.target.tagName === "LABEL") {
            var inputField = event.target.previousElementSibling;
        } else {
            // It is the span inside the target
            var inputField = event.target.parentElement.previousElementSibling;
        }

        if (inputField.checked === true) {
            setCodeCellVisibility(inputField, "visible");
        } else {
            setCodeCellVisibility(inputField, "hidden");
        }
    }


    // Button constructor
    const hideCodeButton = id => `<input class="hidebtn" type="checkbox" id="hidebtn${id}" data-id="${id}"><label title="Toggle cell" for="hidebtn${id}" class="plusminus"><span class="pm_h"></span><span class="pm_v"></span></label>`

    var addHideButton = function () {
        // If a hide button is already added, don't add another
        if (document.querySelector('div.hidecode input') !== null) {
            return;
        }

        // Find the input cells and add a hide button
        document.querySelectorAll('div.input_area').forEach(function (item, index) {
            if (!item.classList.contains("hidecode")) {
                // Skip the cell if it doesn't have a hidecode class
                return;
            }

            const id = codeCellId(index)
            item.setAttribute('id', id);
            // Insert the button just inside the end of the next div
            item.querySelector('div').insertAdjacentHTML('beforeend', hideCodeButton(id))

            // Set up the visibility toggle
            hideLink = document.querySelector(`#${id} div.highlight + input + label`);
            hideLink.addEventListener('click', toggleCodeCellVisibility)
        });
    }


    // Initialize the hide buttos
    var initHiddenCells = function () {
        // Add hide buttons to the cells
        addHideButton();

        // Toggle the code cells that should be hidden
        document.querySelectorAll('div.hidecode input').forEach(function (item) {
            setCodeCellVisibility(item, 'hidden');
            item.checked = true;
        })
    }

    initFunction(initHiddenCells);

</script>

  <!-- Load custom website scripts -->
  <script src="/scRNA-python-workshop/assets/js/scripts.js" async></script>

  <!-- Load custom user CSS and JS  -->
  <script src="/scRNA-python-workshop/assets/custom/custom.js" async></script>
  <link rel="stylesheet" href="/scRNA-python-workshop/assets/custom/custom.css">

  <!-- Update interact links w/ REST param, is defined in includes so we can use templates -->
  
<script>
/**
  * To auto-embed hub URLs in interact links if given in a RESTful fashion
 */

function getJsonFromUrl(url) {
  var query = url.split('?');
  if (query.length < 2) {
    // No queries so just return false
    return false;
  }
  query = query[1];
  // Collect REST params into a dictionary
  var result = {};
  query.split("&").forEach(function(part) {
    var item = part.split("=");
    result[item[0]] = decodeURIComponent(item[1]);
  });
  return result;
}
    
function dict2param(dict) {
    params = Object.keys(dict).map(function(k) {
        return encodeURIComponent(k) + '=' + encodeURIComponent(dict[k])
    });
    return params.join('&')
}

// Parse a Binder URL, converting it to the string needed for JupyterHub
function binder2Jupyterhub(url) {
  newUrl = {};
  parts = url.split('v2/gh/')[1];
  // Grab the base repo information
  repoinfo = parts.split('?')[0];
  var [org, repo, ref] = repoinfo.split('/');
  newUrl['repo'] = ['https://github.com', org, repo].join('/');
  newUrl['branch'] = ref
  // Grab extra parameters passed
  params = getJsonFromUrl(url);
  if (params['filepath'] !== undefined) {
    newUrl['subPath'] = params['filepath']
  }
  return dict2param(newUrl);
}

// Filter out potentially unsafe characters to prevent xss
function safeUrl(url)
{
   return String(encodeURIComponent(url))
            .replace(/&/g, '&amp;')
            .replace(/"/g, '&quot;')
            .replace(/'/g, '&#39;')
            .replace(/</g, '&lt;')
            .replace(/>/g, '&gt;');
}

function addParamToInternalLinks(hub) {
  var links = document.querySelectorAll("a").forEach(function(link) {
    var href = link.href;
    // If the link is an internal link...
    if (href.search("http://localhost:4000") !== -1 || href.startsWith('/') || href.search("127.0.0.1:") !== -1) {
      // Assume we're an internal link, add the hub param to it
      var params = getJsonFromUrl(href);
      if (params !== false) {
        // We have REST params, so append a new one
        params['jupyterhub'] = hub;
      } else {
        // Create the REST params
        params = {'jupyterhub': hub};
      }
      // Update the link
      var newHref = href.split('?')[0] + '?' + dict2param(params);
      link.setAttribute('href', decodeURIComponent(newHref));
    }
  });
  return false;
}


// Update interact links
function updateInteractLink() {
    // hack to make this work since it expects a ? in the URL
    rest = getJsonFromUrl("?" + location.search.substr(1));
    jupyterHubUrl = rest['jupyterhub'];
    var hubType = null;
    var hubUrl = null;
    if (jupyterHubUrl !== undefined) {
      hubType = 'jupyterhub';
      hubUrl = jupyterHubUrl;
    }

    if (hubType !== null) {
      // Sanitize the hubUrl
      hubUrl = safeUrl(hubUrl);

      // Add HTTP text if omitted
      if (hubUrl.indexOf('http') < 0) {hubUrl = 'http://' + hubUrl;}
      var interactButtons = document.querySelectorAll("button.interact-button")
      var lastButton = interactButtons[interactButtons.length-1];
      var link = lastButton.parentElement;

      // If we've already run this, skip the link updating
      if (link.nextElementSibling !== null) {
        return;
      }

      // Update the link and add context div
      var href = link.getAttribute('href');
      if (lastButton.id === 'interact-button-binder') {
        // If binder links exist, we need to re-work them for jupyterhub
        if (hubUrl.indexOf('http%3A%2F%2Flocalhost') > -1) {
          // If localhost, assume we're working from a local Jupyter server and remove `/hub`
          first = [hubUrl, 'git-sync'].join('/')
        } else {
          first = [hubUrl, 'hub', 'user-redirect', 'git-sync'].join('/')
        }
        href = first + '?' + binder2Jupyterhub(href);
      } else {
        // If interact button isn't binderhub, assume it's jupyterhub
        // If JupyterHub links, we only need to replace the hub url
        href = href.replace("", hubUrl);
        if (hubUrl.indexOf('http%3A%2F%2Flocalhost') > -1) {
          // Assume we're working from a local Jupyter server and remove `/hub`
          href = href.replace("/hub/user-redirect", "");
        }
      }
      link.setAttribute('href', decodeURIComponent(href));

      // Add text after interact link saying where we're launching
      hubUrlNoHttp = decodeURIComponent(hubUrl).replace('http://', '').replace('https://', '');
      link.insertAdjacentHTML('afterend', '<div class="interact-context">on ' + hubUrlNoHttp + '</div>');

      // Update internal links so we retain the hub url
      addParamToInternalLinks(hubUrl);
    }
}

runWhenDOMLoaded(updateInteractLink)
document.addEventListener('turbolinks:load', updateInteractLink)
</script>


  <!-- Lunr search code - will only be executed on the /search page -->
  <script src="/scRNA-python-workshop/assets/js/lunr/lunr.min.js" type="text/javascript"></script>
  <script>var initQuery = function() {
  // See if we have a search box
  var searchInput = document.querySelector('input#lunr_search');
  if (searchInput === null) {
    return;
  }

  // Function to parse our lunr cache
  var idx = lunr(function () {
    this.field('title')
    this.field('excerpt')
    this.field('categories')
    this.field('tags')
    this.ref('id')

    this.pipeline.remove(lunr.trimmer)

    for (var item in store) {
      this.add({
        title: store[item].title,
        excerpt: store[item].excerpt,
        categories: store[item].categories,
        tags: store[item].tags,
        id: item
      })
    }
  });

  // Run search upon keyup
  searchInput.addEventListener('keyup', function () {
    var resultdiv = document.querySelector('#results');
    var query = document.querySelector("input#lunr_search").value.toLowerCase();
    var result =
      idx.query(function (q) {
        query.split(lunr.tokenizer.separator).forEach(function (term) {
          q.term(term, { boost: 100 })
          if(query.lastIndexOf(" ") != query.length-1){
            q.term(term, {  usePipeline: false, wildcard: lunr.Query.wildcard.TRAILING, boost: 10 })
          }
          if (term != ""){
            q.term(term, {  usePipeline: false, editDistance: 1, boost: 1 })
          }
        })
      });

      // Empty the results div
      while (resultdiv.firstChild) {
        resultdiv.removeChild(resultdiv.firstChild);
      }

    resultdiv.insertAdjacentHTML('afterbegin', '<p class="results__found">'+result.length+' Result(s) found</p>');
    for (var item in result) {
      var ref = result[item].ref;
      if(store[ref].teaser){
        var searchitem =
          '<div class="list__item">'+
            '<article class="archive__item" itemscope itemtype="https://schema.org/CreativeWork">'+
              '<h2 class="archive__item-title" itemprop="headline">'+
                '<a href="'+store[ref].url+'" rel="permalink">'+store[ref].title+'</a>'+
              '</h2>'+
              '<div class="archive__item-teaser">'+
                '<img src="'+store[ref].teaser+'" alt="">'+
              '</div>'+
              '<p class="archive__item-excerpt" itemprop="description">'+store[ref].excerpt.split(" ").splice(0,20).join(" ")+'...</p>'+
            '</article>'+
          '</div>';
      }
      else{
    	  var searchitem =
          '<div class="list__item">'+
            '<article class="archive__item" itemscope itemtype="https://schema.org/CreativeWork">'+
              '<h2 class="archive__item-title" itemprop="headline">'+
                '<a href="'+store[ref].url+'" rel="permalink">'+store[ref].title+'</a>'+
              '</h2>'+
              '<p class="archive__item-excerpt" itemprop="description">'+store[ref].excerpt.split(" ").splice(0,20).join(" ")+'...</p>'+
            '</article>'+
          '</div>';
      }
      resultdiv.insertAdjacentHTML('beforeend', searchitem);
    }
  });
};

initFunction(initQuery);
</script>
</head>

  <body>
    <!-- .js-show-sidebar shows sidebar by default -->
    <div id="js-textbook" class="c-textbook js-show-sidebar">
      



<nav id="js-sidebar" class="c-textbook__sidebar">
  
  <h2 class="c-sidebar__title">Analysis of single cell RNA-seq data</h2>
  <ul class="c-sidebar__chapters">
    
      
      
        <li><h2 class="c-sidebar__title">Introduction</li>
        
      
      

      
      

      
      
      <li class="c-sidebar__chapter">
        <a class="c-sidebar__entry "
          href="/scRNA-python-workshop/intro/about.html"
        >
          
            1.
          
          About the course
        </a>

        
      </li>

      
    
      
      

      
      

      
      
      <li class="c-sidebar__chapter">
        <a class="c-sidebar__entry "
          href="/scRNA-python-workshop/intro/setup.html"
        >
          
            2.
          
          Setup and installation
        </a>

        
      </li>

      
    
      
      

      
      

      
      
      <li class="c-sidebar__chapter">
        <a class="c-sidebar__entry "
          href="/scRNA-python-workshop/intro/overview.html"
        >
          
            3.
          
          Workflow overview
        </a>

        
      </li>

      
    
      
      
        <li class="c-sidebar__divider"></li>
        
      
      
        <li><h2 class="c-sidebar__title">Data preprocessing</li>
        
      
      

      
      

      
      
      <li class="c-sidebar__chapter">
        <a class="c-sidebar__entry c-sidebar__entry--active"
          href="/scRNA-python-workshop/preprocessing/00-tabula-muris.html"
        >
          
            4.
          
          Tabula muris
        </a>

        
      </li>

      
    
      
      

      
      

      
      
      <li class="c-sidebar__chapter">
        <a class="c-sidebar__entry "
          href="/scRNA-python-workshop/preprocessing/01-basic-qc.html"
        >
          
            5.
          
          Quality control
        </a>

        
      </li>

      
    
      
      

      
      

      
      
      <li class="c-sidebar__chapter">
        <a class="c-sidebar__entry "
          href="/scRNA-python-workshop/preprocessing/02-normalization.html"
        >
          
            6.
          
          Normalization and PCA
        </a>

        
      </li>

      
    
      
      
        <li class="c-sidebar__divider"></li>
        
      
      
        <li><h2 class="c-sidebar__title">Data analysis</li>
        
      
      

      
      

      
      
      <li class="c-sidebar__chapter">
        <a class="c-sidebar__entry "
          href="/scRNA-python-workshop/analysis/03-dimensionality-reduction.html"
        >
          
            7.
          
          Dimensionality reduction
        </a>

        
      </li>

      
    
      
      

      
      

      
      
      <li class="c-sidebar__chapter">
        <a class="c-sidebar__entry "
          href="/scRNA-python-workshop/analysis/04-clustering.html"
        >
          
            8.
          
          Clustering
        </a>

        
      </li>

      
    
      
      

      
      

      
      
      <li class="c-sidebar__chapter">
        <a class="c-sidebar__entry "
          href="/scRNA-python-workshop/analysis/05-diffexp.html"
        >
          
            9.
          
          Differential expression
        </a>

        
      </li>

      
    
      
      
        <li><h2 class="c-sidebar__title">Trajectory inference</li>
        
      
      
        <li class="c-sidebar__divider"></li>
        
      
      
        <li><h2 class="c-sidebar__title">Exploratory analysis & advanced exercises</li>
        
      
      
        <li><h2 class="c-sidebar__title">Advanced exercises</li>
        
      
      
        <li><h2 class="c-sidebar__title">Exploration with cellxgene</li>
        
  </ul>
  <p class="sidebar_footer">Powered by <a href="https://github.com/jupyter/jupyter-book">Jupyter Book</a></p>
</nav>

      
      <!-- Empty sidebar placeholder that we'll auto-fill with javascript -->
      <aside class="sidebar__right">
          <header><h4 class="nav__title"><i class="fa fa-list"></i>   On this page</h4></header>
          <nav class="onthispage">
          </nav>
      </aside>
      
      <main class="c-textbook__page" tabindex="-1">
          <div class="o-wrapper">
            <div class="c-sidebar-toggle">
  <!-- We show the sidebar by default so we use .is-active -->
  <button
    id="js-sidebar-toggle"
    class="hamburger hamburger--arrowalt is-active"
  >
    <span class="hamburger-box">
      <span class="hamburger-inner"></span>
    </span>
    <span class="c-sidebar-toggle__label">Toggle Sidebar</span>
  </button>
</div>

            
<div class="buttons">
<a href="/scRNA-python-workshop/content/preprocessing/00-tabula-muris.ipynb" download>
<button id="interact-button-download" class="interact-button">Download</button>
</a>










<a href="https://mybinder.org/v2/gh/jupyter/jupyter-book/gh-pages?filepath=content%2Fpreprocessing%2F00-tabula-muris.ipynb"><button class="interact-button" id="interact-button-binder"><img class="interact-button-logo" src="/scRNA-python-workshop/assets/images/logo_binder.svg" alt="Interact" />Interact</button></a>


</div>


            <div class="c-textbook__content">
              <h1 id="tabula-muris">Tabula Muris</h1>

<h2 id="introduction">Introduction</h2>

<p>To give you hands-on experience analyzing a single-cell RNASeq dataset from start to finish, we will be using data from <a href="https://tabula-muris.ds.czbiohub.org/">Tabula Muris</a> as an example. The Tabula Muris is a collaborative effort to profile every mouse tissue at a single-cell level. The full dataset includes both high throughput but low-coverage 10X data and lower throughput but high-coverage Smartseq2 data.</p>

<p>For this workshop, we will be using the Smartseq2 data from the mouse brain. This data consists of:</p>
<ol>
  <li>an <strong>expression matrix</strong> where each column corresponds to a gene (or transcript) and each row corresponds to a single cell</li>
  <li>a table of <strong>metadata</strong> describing each cell</li>
</ol>

<h2 id="downloading-the-data">Downloading the data</h2>

<p>You can find the prepared expression matrix and metadata <a href="INSERT LINK">here</a>. You should save these files in the directory <code class="highlighter-rouge">scRNA-couse/content/data/</code>.</p>

<h2 id="reading-the-data">Reading the data</h2>

<p>We can now read in the count matrix from the comma-separated file. Then inspect the resulting dataframe:</p>

<div class="cell code_cell">
  <div class="input_area">
    <div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="n">pd</span> <span class="c">## load in the pandas library for working with dataframes</span>

</code></pre></div>    </div>
  </div>

</div>

<div class="cell code_cell">
  <div class="input_area">
    <div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c">## tell pandas to make a new DataFrame with the contents of `brain_counts.csv`. This might take a minute.</span>
<span class="n">count_dataframe</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s">'../data/brain_counts.csv'</span><span class="p">,</span> <span class="c"># where to find the data</span>
                              <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span> <span class="c"># use the first column to label the rows (the 'index')</span>

<span class="c">## print the first 5 rows of our dataframe</span>
<span class="k">print</span><span class="p">(</span><span class="n">count_dataframe</span><span class="o">.</span><span class="n">head</span><span class="p">())</span> 

</code></pre></div>    </div>
  </div>

  <div class="output_wrapper">
    <div class="output_subarea">
      <div class="output_stream highlighter-rouge"><div class="highlight"><pre class="highlight"><code>                         0610005C13Rik  0610007C21Rik  0610007L01Rik  \
A1.B003290.3_38_F.1.1                0            125             16   
A1.B003728.3_56_F.1.1                0              0              0   
A1.MAA000560.3_10_M.1.1              0            348              0   
A1.MAA000564.3_10_M.1.1              0             41             36   
A1.MAA000923.3_9_M.1.1               0             53              0   

                         0610007N19Rik  0610007P08Rik  0610007P14Rik  \
A1.B003290.3_38_F.1.1                0              0              0   
A1.B003728.3_56_F.1.1                0              0            324   
A1.MAA000560.3_10_M.1.1              0              0              5   
A1.MAA000564.3_10_M.1.1              0              0             24   
A1.MAA000923.3_9_M.1.1               0              0              0   

                         0610007P22Rik  0610008F07Rik  0610009B14Rik  \
A1.B003290.3_38_F.1.1                0              0              0   
A1.B003728.3_56_F.1.1                0              0              0   
A1.MAA000560.3_10_M.1.1              0              0              0   
A1.MAA000564.3_10_M.1.1              0              0              0   
A1.MAA000923.3_9_M.1.1               0              0              0   

                         0610009B22Rik  ...  Zxdb  Zxdc  Zyg11a  Zyg11b  Zyx  \
A1.B003290.3_38_F.1.1                0  ...     0     0       0       0    0   
A1.B003728.3_56_F.1.1                0  ...     0     0       0       0    0   
A1.MAA000560.3_10_M.1.1              0  ...     0     0       0       0    0   
A1.MAA000564.3_10_M.1.1             14  ...     0     0       0       0  125   
A1.MAA000923.3_9_M.1.1               0  ...     0     0       0      81    0   

                         Zzef1  Zzz3  a  l7Rn6  zsGreen_transgene  
A1.B003290.3_38_F.1.1        0     0  0     54                  0  
A1.B003728.3_56_F.1.1        0     0  0      0                  0  
A1.MAA000560.3_10_M.1.1    195     0  0    113                  0  
A1.MAA000564.3_10_M.1.1      0     1  0      0                  0  
A1.MAA000923.3_9_M.1.1       0     0  0      0                  0  

[5 rows x 23433 columns]
</code></pre></div>      </div>
    </div>
  </div>
</div>

<h3 id="exercise">Exercise</h3>

<p>What do the column names represent? What do the row names represent? How many cells and genes are in this dataset?</p>

<p>Hint: need some help? Try <code class="highlighter-rouge">help(pd.DataFrame.shape)</code>.</p>

<p>
<details>
<summary><h3>Solution</h3></summary>
The column names represent genes. 
The row names represent unique cell identifiers that were assigned by the authors of the dataset.
<p>
We can find out how many genes and cells are in the dataset by asking for its <code>shape</code>:  <br />
<code>print(count_dataframe.shape)</code><br />
gives:<br />
<code>(3401, 23433)</code>
<p>
which represents <code>(N rows, N columns)</code>.
&lt;/details&gt;



## Reading the metadata

The authors have also provided metadata describing each cell. This metadata is stored in a separate file, `brain_metadata.csv`. We can load it into a dataframe and inspect it, just like we did for the count data. 

### Exercise

Load the metadata from the csv file into a pandas dataframe called `metadata_dataframe`. Does it have the same dimensions and index as the `counts_dataframe`?

<p>
<details>
<summary><h3>Solution</h3></summary>
First, we need to load in the metadata.<br />
<code>metadata_dataframe = pd.read_csv('../data/brain_metadata.csv', # where to find the data
                           index_col=0) # use the first column as the index</code>
    
<p>
Let's take a peak at the resulting dataframe to make sure it looks correct. <br />
<code>print(metadata_dataframe.shape)</code><br />
<code>&gt;&gt;&gt; (3401, 5)</code><br />
We have 5 columns of information about 3,401 cells. Sounds reasonable.
    
<p>
Let's take a closer look and inspect the first few rows:<br />
<code>print(metadata_dataframe.head())</code><br />
<code>&gt;&gt;&gt;                        cell_ontology_class    subtissue mouse.sex mouse.id  \
cell                                                                          
A1.B003290.3_38_F.1.1             astrocyte     Striatum         F   3_38_F   
A1.B003728.3_56_F.1.1             astrocyte     Striatum         F   3_56_F   
A1.MAA000560.3_10_M.1.1     oligodendrocyte       Cortex         M   3_10_M   
A1.MAA000564.3_10_M.1.1    endothelial cell     Striatum         M   3_10_M   
A1.MAA000923.3_9_M.1.1            astrocyte  Hippocampus         M    3_9_M   
                        plate.barcode  
cell                                   
A1.B003290.3_38_F.1.1         B003290  
A1.B003728.3_56_F.1.1         B003728  
A1.MAA000560.3_10_M.1.1     MAA000560  
A1.MAA000564.3_10_M.1.1     MAA000564  
A1.MAA000923.3_9_M.1.1      MAA000923</code>
&lt;/details&gt;



### Exercise

To get a sense for what is in this dataset, let's look at the summary of each metadata column. 

1. How many cells from each `subtissue` label are in this dataset? <br />
_Hint: try running `help(pd.value_counts())` to get started_  
<p>
2. Using a `for` loop, repeat this counting procedure to summarize each of the metadata columns.  <br />
_Hint: you can access all the column names in the dataframe with `columns = metadata_dataframe.columns.values`_  


<p>
<details>
<summary><h3>Solution</h3></summary>
1. We can count the number of times each value appears in a column like this:<br />
    <code>print(pd.value_counts(metadata_dataframe['subtissue']))</code><br />
    <code>&gt;&gt;&gt;        
Cortex         1149
Hippocampus     976
Striatum        723
Cerebellum      553
Name: subtissue, dtype: int64</code>
    <p>
2. To repeat this for each column in the dataframe, we can use a <code>for</code> loop:<br />
<code>for column in metadata_dataframe.columns.values:
    print(pd.value_counts(metadata_dataframe[column]))</code>
&lt;/details&gt;



## Building an AnnData object



We now have two dataframes, containing the counts and metadata from the Tabula Muris brain dataset. To keep these organized, we'll use a **data structure** called **AnnData**. AnnData stands for "annotated data," and is the standard format used by the analysis library, **SCANPY**. 

AnnData uses some generalized vocabulary to describe cells and genes: they refer to cells as **observations** and genes as **variables**. This data structure has four areas where we can store information:  

<img src="../figures/anndata.svg" alt="Drawing" style="width: 500px;" />

`AnnData.X` stores the count matrix  
`AnnData.obs` stores metadata about the observations (cells)  
`AnnData.var` stores metadata about the variables (genes)  
`AnnData.uns` stores any additional, unstructured information we decide to attach later  



Here, we have a count matrix and metadata that describes each cell, so we will use the `.X` and `.obs` portions of the AnnData structure. 



<div class="cell code_cell">
                          <div class="input_area">
                            <div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="n">sc</span> <span class="c"># import the scanpy library that tells Python how an AnnData data structure works</span>

</code></pre></div>                            </div>
                          </div>

                        </div>



<div class="cell code_cell">
                          <div class="input_area">
                            <div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c"># help(sc.AnnData)</span>
<span class="n">adata</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">AnnData</span><span class="p">(</span><span class="n">X</span> <span class="o">=</span> <span class="n">count_dataframe</span><span class="p">,</span> <span class="n">obs</span> <span class="o">=</span> <span class="n">metadata_dataframe</span><span class="p">)</span>
<span class="k">print</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>

</code></pre></div>                            </div>
                          </div>

                          <div class="output_wrapper">
                            <div class="output_subarea">
                              <div class="output_stream highlighter-rouge"><div class="highlight"><pre class="highlight"><code>AnnData object with n_obs × n_vars = 3401 × 23433 
    obs: 'cell_ontology_class', 'subtissue', 'mouse.sex', 'mouse.id', 'plate.barcode'
</code></pre></div>                              </div>
                            </div>
                          </div>
                        </div>



### Labeling spike-ins
Because this is smartseq2 data, we may have spike-ins. These gene names start with `ERCC`. We can label them in `adata.var` as a gene annotation.



<div class="cell code_cell">
                          <div class="input_area">
                            <div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">is_spike_in</span> <span class="o">=</span> <span class="p">{}</span>
<span class="n">number_of_spike_ins</span> <span class="o">=</span> <span class="mi">0</span>

<span class="k">for</span> <span class="n">gene_name</span> <span class="ow">in</span> <span class="n">adata</span><span class="o">.</span><span class="n">var_names</span><span class="p">:</span>
    <span class="k">if</span> <span class="s">'ERCC'</span> <span class="ow">in</span> <span class="n">gene_name</span><span class="p">:</span>
        <span class="n">is_spike_in</span><span class="p">[</span><span class="n">gene_name</span><span class="p">]</span> <span class="o">=</span> <span class="bp">True</span> <span class="c"># record that we found a spike-in</span>
        <span class="n">number_of_spike_ins</span> <span class="o">+=</span> <span class="mi">1</span> <span class="c"># bump the counter</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">is_spike_in</span><span class="p">[</span><span class="n">gene_name</span><span class="p">]</span> <span class="o">=</span> <span class="bp">False</span> <span class="c"># record that this was not a spike-in</span>
        
<span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s">'ERCC'</span><span class="p">]</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">is_spike_in</span><span class="p">)</span> <span class="c"># because the index of adata.var and the keys of is_spike_in match, anndata will take care of matching them up</span>
<span class="k">print</span><span class="p">(</span><span class="s">'found this many spike ins: '</span><span class="p">,</span> <span class="n">number_of_spike_ins</span><span class="p">)</span>

</code></pre></div>                            </div>
                          </div>

                          <div class="output_wrapper">
                            <div class="output_subarea">
                              <div class="output_stream highlighter-rouge"><div class="highlight"><pre class="highlight"><code>found this many spike ins:  92
</code></pre></div>                              </div>
                            </div>
                          </div>
                        </div>



Now that we've finished building our AnnData object, we can save it in a file for later use like so:



<div class="cell code_cell">
                          <div class="input_area">
                            <div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">adata</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">'../data/brain_raw.h5ad'</span><span class="p">)</span> <span class="c">## the h5ad extension is AnnData-specific</span>

</code></pre></div>                            </div>
                          </div>

                        </div>

</p></details></p></p></p></p></details></p></p></p></details></p>

              <nav class="c-page__nav">
  
    
    <a id="js-page__nav__prev" class="c-page__nav__prev" href="/scRNA-python-workshop/intro/overview.html">
      〈 <span class="u-margin-right-tiny"></span> Workflow overview
    </a>
  

  
    
    <a id="js-page__nav__next" class="c-page__nav__next" href="/scRNA-python-workshop/preprocessing/01-basic-qc.html">
      Quality control <span class="u-margin-right-tiny"></span> 〉
    </a>
  
</nav>

            </div>
          </div>
        </div>
      </main>
    </div>

  </body>
</html>
