<!DOCTYPE html>
<html lang="en">
  

  <head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width,minimum-scale=1">

  <title>Quality control</title>
  <meta name="description" content="Introduction">

  <link rel="canonical" href="http://localhost:4000/scRNA-python-workshop/preprocessing/01-basic-qc.html">
  <link rel="alternate" type="application/rss+xml" title="Analysis of single cell RNA-seq data" href="http://localhost:4000/scRNA-python-workshop/feed.xml">

  <meta property="og:url"         content="http://localhost:4000/scRNA-python-workshop/preprocessing/01-basic-qc.html" />
<meta property="og:type"        content="article" />
<meta property="og:title"       content="Quality control" />
<meta property="og:description" content="Introduction" />
<meta property="og:image"       content="" />


  <script type="application/ld+json">
  {
  "@context": "http://schema.org",
  "@type": "NewsArticle",
  "mainEntityOfPage":
    "http://localhost:4000/scRNA-python-workshop/preprocessing/01-basic-qc.html",
  "headline":
    "Quality control",
  "datePublished":
    "2019-09-23T08:46:14-07:00",
  "dateModified":
    "2019-09-23T08:46:14-07:00",
  "description":
    "Introduction",
  "author": {
    "@type": "Person",
    "name": "Vladimir Kiselev & the Hemberg lab; adapted by Sidney Bell"
  },
  "publisher": {
    "@type": "Organization",
    "name": "Data 100 at UC Berkeley",
    "logo": {
      "@type": "ImageObject",
      "url": "http://localhost:4000/scRNA-python-workshop",
      "width": 60,
      "height": 60
    }
  },
  "image": {
    "@type": "ImageObject",
    "url": "http://localhost:4000/scRNA-python-workshop",
    "height": 60,
    "width": 60
  }
}

  </script>
  <link rel="stylesheet" href="/scRNA-python-workshop/assets/css/styles.css">
  <link rel="stylesheet" href="https://stackpath.bootstrapcdn.com/font-awesome/4.7.0/css/font-awesome.min.css ">
  <link rel="apple-touch-icon" sizes="57x57" href="/apple-touch-icon-57x57.png">
  <link rel="apple-touch-icon" sizes="60x60" href="/apple-touch-icon-60x60.png">
  <link rel="apple-touch-icon" sizes="72x72" href="/apple-touch-icon-72x72.png">
  <link rel="apple-touch-icon" sizes="76x76" href="/apple-touch-icon-76x76.png">
  <link rel="apple-touch-icon" sizes="114x114" href="/apple-touch-icon-114x114.png">
  <link rel="apple-touch-icon" sizes="120x120" href="/apple-touch-icon-120x120.png">
  <link rel="apple-touch-icon" sizes="144x144" href="/apple-touch-icon-144x144.png">
  <link rel="apple-touch-icon" sizes="152x152" href="/apple-touch-icon-152x152.png">
  <link rel="apple-touch-icon" sizes="180x180" href="/apple-touch-icon-180x180.png">

  <!-- <link rel="manifest" href="/manifest.json"> -->
  <!-- <link rel="mask-icon" href="/safari-pinned-tab.svg" color="#efae0a"> -->
  <meta name="msapplication-TileColor" content="#da532c">
  <meta name="msapplication-TileImage" content="/mstile-144x144.png">
  <meta name="theme-color" content="#233947">

  <!-- Favicon -->
  <link rel="shortcut icon" type="image/x-icon" href="/scRNA-python-workshop/images/logo/favicon.ico">

  <!-- MathJax Config -->
  <!-- Allow inline math using $ and automatically break long math lines -->
<script type="text/x-mathjax-config">
MathJax.Hub.Config({
    tex2jax: {
        inlineMath: [ ['$','$'], ["\\(","\\)"] ],
        displayMath: [ ['$$','$$'], ["\\[","\\]"] ],
        processEscapes: true,
        processEnvironments: true
    },
    CommonHTML: {
        linebreaks: {
            automatic: true,
        },
    },
});
</script>
<script src='https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js?config=TeX-AMS_CHTML' async></script>

  <!-- DOM updating function -->
  <script>
const runWhenDOMLoaded = cb => {
  if (document.readyState != 'loading') {
    cb()
  } else if (document.addEventListener) {
    document.addEventListener('DOMContentLoaded', cb)
  } else {
    document.attachEvent('onreadystatechange', function() {
      if (document.readyState == 'complete') cb()
    })
  }
}

// Helper function to init things quickly
initFunction = function(myfunc) {
  runWhenDOMLoaded(myfunc);
  document.addEventListener('turbolinks:load', myfunc);
};
</script>

  <!-- Define some javascript variables that will be useful in other javascript -->
  <script>
    const site_basename = '/scRNA-python-workshop';
  </script>

  <!-- Add AnchorJS to let headers be linked -->
  <script src="/scRNA-python-workshop/assets/js/anchor.min.js"  type="text/javascript"></script>
  <script>

initFunction(function () {
    anchors.add("main h1, main h2, main h3, main h4")
});

</script>

  <!-- Include Turbolinks to make page loads fast -->
  <!-- https://github.com/turbolinks/turbolinks -->
  <script src="/scRNA-python-workshop/assets/js/turbolinks.js" async></script>
  <meta name="turbolinks-cache-control" content="no-cache">

  <!-- Load nbinteract for widgets -->
  

  <!-- Load Thebelab for interactive widgets -->
  <!-- Include Thebelab for interactive code if it's enabled -->



  <!-- Load the auto-generating TOC -->
  <script src="/scRNA-python-workshop/assets/js/tocbot.min.js"  type="text/javascript"></script>
  <script>
var initToc = function () {
  tocbot.init({
    tocSelector: 'nav.onthispage',
    contentSelector: '.c-textbook__content',
    headingSelector: 'h2, h3',
    orderedList: false,
    collapseDepth: 6,
    listClass: 'toc__menu',
    activeListItemClass: "",  // Not using
    activeLinkClass: "", // Not using
  });
  tocbot.refresh();
}
initFunction(initToc);
</script>

  <!-- Google analytics -->
  <script src="/scRNA-python-workshop/assets/js/ga.js" async></script>

  <!-- Clipboard copy button -->
  <script src="https://cdn.jsdelivr.net/npm/clipboard@2/dist/clipboard.min.js" async></script>

  <!-- Load JS that depends on site variables -->
  <script>
/**
 * Set up copy/paste for code blocks
 */
const codeCellId = index => `codecell${index}`

const clipboardButton = id =>
  `<a id="copy-button-${id}" class="btn copybtn o-tooltip--left" data-tooltip="Copy" data-clipboard-target="#${id}">
    <img src="/scRNA-python-workshop/assets/images/copy-button.svg" alt="Copy to clipboard">
  </a>`

// Clears selected text since ClipboardJS will select the text when copying
const clearSelection = () => {
  if (window.getSelection) {
    window.getSelection().removeAllRanges()
  } else if (document.selection) {
    document.selection.empty()
  }
}

// Changes tooltip text for two seconds, then changes it back
const temporarilyChangeTooltip = (el, newText) => {
  const oldText = el.getAttribute('data-tooltip')
  el.setAttribute('data-tooltip', newText)
  setTimeout(() => el.setAttribute('data-tooltip', oldText), 2000)
}

const addCopyButtonToCodeCells = () => {
  // If ClipboardJS hasn't loaded, wait a bit and try again. This
  // happens because we load ClipboardJS asynchronously.
  if (window.ClipboardJS === undefined) {
    setTimeout(addCopyButtonToCodeCells, 250)
    return
  }

  const codeCells = document.querySelectorAll('div.c-textbook__content > div.highlighter-rouge > div.highlight > pre, div.input_area pre')
  codeCells.forEach((codeCell, index) => {
    const id = codeCellId(index)
    codeCell.setAttribute('id', id)
    if (document.getElementById("copy-button" + id) == null) {
      codeCell.insertAdjacentHTML('afterend', clipboardButton(id));
    }
  })

  const clipboard = new ClipboardJS('.copybtn')
  clipboard.on('success', event => {
    clearSelection()
    temporarilyChangeTooltip(event.trigger, 'Copied!')
  })

  clipboard.on('error', event => {
    temporarilyChangeTooltip(event.trigger, 'Failed to copy')
  })

  // Get rid of clipboard before the next page visit to avoid memory leak
  document.addEventListener('turbolinks:before-visit', () =>
    clipboard.destroy()
  )
}

initFunction(addCopyButtonToCodeCells);
</script>


  <!-- Hide cell code -->
  <script>
    /**
    Add buttons to hide code cells
    */


    var setCodeCellVisibility = function (inputField, kind) {
        // Update the image and class for hidden
        var id = inputField.getAttribute('data-id');
        var codeCell = document.querySelector(`#${id} div.highlight`);

        if (kind === "visible") {
            codeCell.classList.remove('hidden');
            inputField.checked = true;
        } else {
            codeCell.classList.add('hidden');
            inputField.checked = false;
        }
    }

    var toggleCodeCellVisibility = function (event) {
        // The label is clicked, and now we decide what to do based on the input field's clicked status
        if (event.target.tagName === "LABEL") {
            var inputField = event.target.previousElementSibling;
        } else {
            // It is the span inside the target
            var inputField = event.target.parentElement.previousElementSibling;
        }

        if (inputField.checked === true) {
            setCodeCellVisibility(inputField, "visible");
        } else {
            setCodeCellVisibility(inputField, "hidden");
        }
    }


    // Button constructor
    const hideCodeButton = id => `<input class="hidebtn" type="checkbox" id="hidebtn${id}" data-id="${id}"><label title="Toggle cell" for="hidebtn${id}" class="plusminus"><span class="pm_h"></span><span class="pm_v"></span></label>`

    var addHideButton = function () {
        // If a hide button is already added, don't add another
        if (document.querySelector('div.hidecode input') !== null) {
            return;
        }

        // Find the input cells and add a hide button
        document.querySelectorAll('div.input_area').forEach(function (item, index) {
            if (!item.classList.contains("hidecode")) {
                // Skip the cell if it doesn't have a hidecode class
                return;
            }

            const id = codeCellId(index)
            item.setAttribute('id', id);
            // Insert the button just inside the end of the next div
            item.querySelector('div').insertAdjacentHTML('beforeend', hideCodeButton(id))

            // Set up the visibility toggle
            hideLink = document.querySelector(`#${id} div.highlight + input + label`);
            hideLink.addEventListener('click', toggleCodeCellVisibility)
        });
    }


    // Initialize the hide buttos
    var initHiddenCells = function () {
        // Add hide buttons to the cells
        addHideButton();

        // Toggle the code cells that should be hidden
        document.querySelectorAll('div.hidecode input').forEach(function (item) {
            setCodeCellVisibility(item, 'hidden');
            item.checked = true;
        })
    }

    initFunction(initHiddenCells);

</script>

  <!-- Load custom website scripts -->
  <script src="/scRNA-python-workshop/assets/js/scripts.js" async></script>

  <!-- Load custom user CSS and JS  -->
  <script src="/scRNA-python-workshop/assets/custom/custom.js" async></script>
  <link rel="stylesheet" href="/scRNA-python-workshop/assets/custom/custom.css">

  <!-- Update interact links w/ REST param, is defined in includes so we can use templates -->
  
<script>
/**
  * To auto-embed hub URLs in interact links if given in a RESTful fashion
 */

function getJsonFromUrl(url) {
  var query = url.split('?');
  if (query.length < 2) {
    // No queries so just return false
    return false;
  }
  query = query[1];
  // Collect REST params into a dictionary
  var result = {};
  query.split("&").forEach(function(part) {
    var item = part.split("=");
    result[item[0]] = decodeURIComponent(item[1]);
  });
  return result;
}
    
function dict2param(dict) {
    params = Object.keys(dict).map(function(k) {
        return encodeURIComponent(k) + '=' + encodeURIComponent(dict[k])
    });
    return params.join('&')
}

// Parse a Binder URL, converting it to the string needed for JupyterHub
function binder2Jupyterhub(url) {
  newUrl = {};
  parts = url.split('v2/gh/')[1];
  // Grab the base repo information
  repoinfo = parts.split('?')[0];
  var [org, repo, ref] = repoinfo.split('/');
  newUrl['repo'] = ['https://github.com', org, repo].join('/');
  newUrl['branch'] = ref
  // Grab extra parameters passed
  params = getJsonFromUrl(url);
  if (params['filepath'] !== undefined) {
    newUrl['subPath'] = params['filepath']
  }
  return dict2param(newUrl);
}

// Filter out potentially unsafe characters to prevent xss
function safeUrl(url)
{
   return String(encodeURIComponent(url))
            .replace(/&/g, '&amp;')
            .replace(/"/g, '&quot;')
            .replace(/'/g, '&#39;')
            .replace(/</g, '&lt;')
            .replace(/>/g, '&gt;');
}

function addParamToInternalLinks(hub) {
  var links = document.querySelectorAll("a").forEach(function(link) {
    var href = link.href;
    // If the link is an internal link...
    if (href.search("http://localhost:4000") !== -1 || href.startsWith('/') || href.search("127.0.0.1:") !== -1) {
      // Assume we're an internal link, add the hub param to it
      var params = getJsonFromUrl(href);
      if (params !== false) {
        // We have REST params, so append a new one
        params['jupyterhub'] = hub;
      } else {
        // Create the REST params
        params = {'jupyterhub': hub};
      }
      // Update the link
      var newHref = href.split('?')[0] + '?' + dict2param(params);
      link.setAttribute('href', decodeURIComponent(newHref));
    }
  });
  return false;
}


// Update interact links
function updateInteractLink() {
    // hack to make this work since it expects a ? in the URL
    rest = getJsonFromUrl("?" + location.search.substr(1));
    jupyterHubUrl = rest['jupyterhub'];
    var hubType = null;
    var hubUrl = null;
    if (jupyterHubUrl !== undefined) {
      hubType = 'jupyterhub';
      hubUrl = jupyterHubUrl;
    }

    if (hubType !== null) {
      // Sanitize the hubUrl
      hubUrl = safeUrl(hubUrl);

      // Add HTTP text if omitted
      if (hubUrl.indexOf('http') < 0) {hubUrl = 'http://' + hubUrl;}
      var interactButtons = document.querySelectorAll("button.interact-button")
      var lastButton = interactButtons[interactButtons.length-1];
      var link = lastButton.parentElement;

      // If we've already run this, skip the link updating
      if (link.nextElementSibling !== null) {
        return;
      }

      // Update the link and add context div
      var href = link.getAttribute('href');
      if (lastButton.id === 'interact-button-binder') {
        // If binder links exist, we need to re-work them for jupyterhub
        if (hubUrl.indexOf('http%3A%2F%2Flocalhost') > -1) {
          // If localhost, assume we're working from a local Jupyter server and remove `/hub`
          first = [hubUrl, 'git-sync'].join('/')
        } else {
          first = [hubUrl, 'hub', 'user-redirect', 'git-sync'].join('/')
        }
        href = first + '?' + binder2Jupyterhub(href);
      } else {
        // If interact button isn't binderhub, assume it's jupyterhub
        // If JupyterHub links, we only need to replace the hub url
        href = href.replace("", hubUrl);
        if (hubUrl.indexOf('http%3A%2F%2Flocalhost') > -1) {
          // Assume we're working from a local Jupyter server and remove `/hub`
          href = href.replace("/hub/user-redirect", "");
        }
      }
      link.setAttribute('href', decodeURIComponent(href));

      // Add text after interact link saying where we're launching
      hubUrlNoHttp = decodeURIComponent(hubUrl).replace('http://', '').replace('https://', '');
      link.insertAdjacentHTML('afterend', '<div class="interact-context">on ' + hubUrlNoHttp + '</div>');

      // Update internal links so we retain the hub url
      addParamToInternalLinks(hubUrl);
    }
}

runWhenDOMLoaded(updateInteractLink)
document.addEventListener('turbolinks:load', updateInteractLink)
</script>


  <!-- Lunr search code - will only be executed on the /search page -->
  <script src="/scRNA-python-workshop/assets/js/lunr/lunr.min.js" type="text/javascript"></script>
  <script>var initQuery = function() {
  // See if we have a search box
  var searchInput = document.querySelector('input#lunr_search');
  if (searchInput === null) {
    return;
  }

  // Function to parse our lunr cache
  var idx = lunr(function () {
    this.field('title')
    this.field('excerpt')
    this.field('categories')
    this.field('tags')
    this.ref('id')

    this.pipeline.remove(lunr.trimmer)

    for (var item in store) {
      this.add({
        title: store[item].title,
        excerpt: store[item].excerpt,
        categories: store[item].categories,
        tags: store[item].tags,
        id: item
      })
    }
  });

  // Run search upon keyup
  searchInput.addEventListener('keyup', function () {
    var resultdiv = document.querySelector('#results');
    var query = document.querySelector("input#lunr_search").value.toLowerCase();
    var result =
      idx.query(function (q) {
        query.split(lunr.tokenizer.separator).forEach(function (term) {
          q.term(term, { boost: 100 })
          if(query.lastIndexOf(" ") != query.length-1){
            q.term(term, {  usePipeline: false, wildcard: lunr.Query.wildcard.TRAILING, boost: 10 })
          }
          if (term != ""){
            q.term(term, {  usePipeline: false, editDistance: 1, boost: 1 })
          }
        })
      });

      // Empty the results div
      while (resultdiv.firstChild) {
        resultdiv.removeChild(resultdiv.firstChild);
      }

    resultdiv.insertAdjacentHTML('afterbegin', '<p class="results__found">'+result.length+' Result(s) found</p>');
    for (var item in result) {
      var ref = result[item].ref;
      if(store[ref].teaser){
        var searchitem =
          '<div class="list__item">'+
            '<article class="archive__item" itemscope itemtype="https://schema.org/CreativeWork">'+
              '<h2 class="archive__item-title" itemprop="headline">'+
                '<a href="'+store[ref].url+'" rel="permalink">'+store[ref].title+'</a>'+
              '</h2>'+
              '<div class="archive__item-teaser">'+
                '<img src="'+store[ref].teaser+'" alt="">'+
              '</div>'+
              '<p class="archive__item-excerpt" itemprop="description">'+store[ref].excerpt.split(" ").splice(0,20).join(" ")+'...</p>'+
            '</article>'+
          '</div>';
      }
      else{
    	  var searchitem =
          '<div class="list__item">'+
            '<article class="archive__item" itemscope itemtype="https://schema.org/CreativeWork">'+
              '<h2 class="archive__item-title" itemprop="headline">'+
                '<a href="'+store[ref].url+'" rel="permalink">'+store[ref].title+'</a>'+
              '</h2>'+
              '<p class="archive__item-excerpt" itemprop="description">'+store[ref].excerpt.split(" ").splice(0,20).join(" ")+'...</p>'+
            '</article>'+
          '</div>';
      }
      resultdiv.insertAdjacentHTML('beforeend', searchitem);
    }
  });
};

initFunction(initQuery);
</script>
</head>

  <body>
    <!-- .js-show-sidebar shows sidebar by default -->
    <div id="js-textbook" class="c-textbook js-show-sidebar">
      



<nav id="js-sidebar" class="c-textbook__sidebar">
  
  <h2 class="c-sidebar__title">Analysis of single cell RNA-seq data</h2>
  <ul class="c-sidebar__chapters">
    
      
      
        <li><h2 class="c-sidebar__title">Introduction</li>
        
      
      

      
      

      
      
      <li class="c-sidebar__chapter">
        <a class="c-sidebar__entry "
          href="/scRNA-python-workshop/intro/about.html"
        >
          
            1.
          
          About the course
        </a>

        
      </li>

      
    
      
      

      
      

      
      
      <li class="c-sidebar__chapter">
        <a class="c-sidebar__entry "
          href="/scRNA-python-workshop/intro/setup.html"
        >
          
            2.
          
          Setup and installation
        </a>

        
      </li>

      
    
      
      

      
      

      
      
      <li class="c-sidebar__chapter">
        <a class="c-sidebar__entry "
          href="/scRNA-python-workshop/intro/overview.html"
        >
          
            3.
          
          Workflow overview
        </a>

        
      </li>

      
    
      
      
        <li class="c-sidebar__divider"></li>
        
      
      
        <li><h2 class="c-sidebar__title">Data preprocessing</li>
        
      
      

      
      

      
      
      <li class="c-sidebar__chapter">
        <a class="c-sidebar__entry "
          href="/scRNA-python-workshop/preprocessing/00-tabula-muris.html"
        >
          
            4.
          
          Tabula muris
        </a>

        
      </li>

      
    
      
      

      
      

      
      
      <li class="c-sidebar__chapter">
        <a class="c-sidebar__entry c-sidebar__entry--active"
          href="/scRNA-python-workshop/preprocessing/01-basic-qc.html"
        >
          
            5.
          
          Quality control
        </a>

        
      </li>

      
    
      
      

      
      

      
      
      <li class="c-sidebar__chapter">
        <a class="c-sidebar__entry "
          href="/scRNA-python-workshop/preprocessing/02-normalization.html"
        >
          
            6.
          
          Normalization and PCA
        </a>

        
      </li>

      
    
      
      
        <li class="c-sidebar__divider"></li>
        
      
      
        <li><h2 class="c-sidebar__title">Data analysis</li>
        
      
      

      
      

      
      
      <li class="c-sidebar__chapter">
        <a class="c-sidebar__entry "
          href="/scRNA-python-workshop/analysis/03-dimensionality-reduction.html"
        >
          
            7.
          
          Dimensionality reduction
        </a>

        
      </li>

      
    
      
      

      
      

      
      
      <li class="c-sidebar__chapter">
        <a class="c-sidebar__entry "
          href="/scRNA-python-workshop/analysis/04-clustering.html"
        >
          
            8.
          
          Clustering
        </a>

        
      </li>

      
    
      
      

      
      

      
      
      <li class="c-sidebar__chapter">
        <a class="c-sidebar__entry "
          href="/scRNA-python-workshop/analysis/05-diffexp.html"
        >
          
            9.
          
          Differential expression
        </a>

        
      </li>

      
    
      
      
        <li><h2 class="c-sidebar__title">Trajectory inference</li>
        
      
      
        <li class="c-sidebar__divider"></li>
        
      
      
        <li><h2 class="c-sidebar__title">Exploratory analysis & advanced exercises</li>
        
      
      
        <li><h2 class="c-sidebar__title">Advanced exercises</li>
        
      
      
        <li><h2 class="c-sidebar__title">Exploration with cellxgene</li>
        
  </ul>
  <p class="sidebar_footer">Powered by <a href="https://github.com/jupyter/jupyter-book">Jupyter Book</a></p>
</nav>

      
      <!-- Empty sidebar placeholder that we'll auto-fill with javascript -->
      <aside class="sidebar__right">
          <header><h4 class="nav__title"><i class="fa fa-list"></i>   On this page</h4></header>
          <nav class="onthispage">
          </nav>
      </aside>
      
      <main class="c-textbook__page" tabindex="-1">
          <div class="o-wrapper">
            <div class="c-sidebar-toggle">
  <!-- We show the sidebar by default so we use .is-active -->
  <button
    id="js-sidebar-toggle"
    class="hamburger hamburger--arrowalt is-active"
  >
    <span class="hamburger-box">
      <span class="hamburger-inner"></span>
    </span>
    <span class="c-sidebar-toggle__label">Toggle Sidebar</span>
  </button>
</div>

            
<div class="buttons">
<a href="/scRNA-python-workshop/content/preprocessing/01-basic-qc.ipynb" download>
<button id="interact-button-download" class="interact-button">Download</button>
</a>










<a href="https://mybinder.org/v2/gh/jupyter/jupyter-book/gh-pages?filepath=content%2Fpreprocessing%2F01-basic-qc.ipynb"><button class="interact-button" id="interact-button-binder"><img class="interact-button-logo" src="/scRNA-python-workshop/assets/images/logo_binder.svg" alt="Interact" />Interact</button></a>


</div>


            <div class="c-textbook__content">
              <h1 id="introduction">Introduction</h1>

<p>Once we have our expression matrix, it should be examined to remove poor quality cells which were not detected in the initial processing of the raw reads. Failure to remove low quality cells at this
stage may add technical noise which has the potential to obscure
the biological signals of interest in the downstream analysis.</p>

<p>Since there is currently no standard method for performing scRNAseq, the expected values for the various QC measures that will be presented here can vary substantially from experiment to experiment. Thus, to perform QC we will be looking for cells which are outliers with respect to the rest of the dataset rather than comparing to independent quality standards. Consequently, care should be taken when comparing quality metrics across datasets collected using different protocols.</p>

<h1 id="dataset">Dataset</h1>

<p>We’ll continue working with the Tabula Muris brain data that we prepared earlier:</p>

<div class="cell code_cell">
  <div class="input_area">
    <div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="n">sc</span> <span class="c"># import scanpy to handle our AnnData </span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="n">pd</span> <span class="c"># import pandas to handle dataframes</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="n">plt</span> <span class="c"># import matplotlib to visualize our qc metrics</span>

<span class="c"># magic incantation to help matplotlib work with our jupyter notebook</span>
<span class="o">%</span><span class="n">matplotlib</span> <span class="n">inline</span> 

</code></pre></div>    </div>
  </div>

</div>

<div class="cell code_cell">
  <div class="input_area">
    <div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">adata</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s">'../data/brain_raw.h5ad'</span><span class="p">)</span>

</code></pre></div>    </div>
  </div>

</div>

<h2 id="computing-quality-control-metrics">Computing quality control metrics</h2>
<p>We’ll compute quality metrics and then filter cells and genes accordingly.</p>

<p>The <code class="highlighter-rouge">calculate_qc_metrics</code> function returns two dataframes: one containing quality control metrics about cells, and one containing metrics about genes. This function is housed in the ‘preprocessing’ portion of the SCANPY library, which you can read more about <a href="https://icb-scanpy.readthedocs-hosted.com/en/stable/api/index.html#module-scanpy.pp">here</a>.</p>

<div class="cell code_cell">
  <div class="input_area">
    <div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">qc</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">calculate_qc_metrics</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">qc_vars</span> <span class="o">=</span> <span class="p">[</span><span class="s">'ERCC'</span><span class="p">])</span><span class="c"># this returns a tuple of (cell_qc_dataframe, gene_qc_dataframe)</span>
                                 <span class="c"># ask for the percentage of reads from spike ins</span>
                                
<span class="n">cell_qc_dataframe</span> <span class="o">=</span> <span class="n">qc</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="n">gene_qc_dataframe</span> <span class="o">=</span> <span class="n">qc</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>

<span class="k">print</span><span class="p">(</span><span class="s">'This is the cell quality control dataframe:'</span><span class="p">)</span>
<span class="k">print</span><span class="p">(</span><span class="n">cell_qc_dataframe</span><span class="o">.</span><span class="n">head</span><span class="p">())</span>

<span class="k">print</span><span class="p">(</span><span class="s">'</span><span class="se">\n\n</span><span class="s">This is the gene quality control dataframe'</span><span class="p">)</span>
<span class="k">print</span><span class="p">(</span><span class="n">gene_qc_dataframe</span><span class="p">)</span>

</code></pre></div>    </div>
  </div>

  <div class="output_wrapper">
    <div class="output_subarea">
      <div class="output_stream highlighter-rouge"><div class="highlight"><pre class="highlight"><code>This is the cell quality control dataframe:
                         n_genes_by_counts  log1p_n_genes_by_counts  \
index                                                                 
A1.B003290.3_38_F.1.1                 3359                 8.119696   
A1.B003728.3_56_F.1.1                 1718                 7.449498   
A1.MAA000560.3_10_M.1.1               3910                 8.271548   
A1.MAA000564.3_10_M.1.1               4352                 8.378621   
A1.MAA000923.3_9_M.1.1                2248                 7.718241   

                         total_counts  log1p_total_counts  \
index                                                       
A1.B003290.3_38_F.1.1        390075.0           12.874097   
A1.B003728.3_56_F.1.1        776439.0           13.562474   
A1.MAA000560.3_10_M.1.1     1616087.0           14.295519   
A1.MAA000564.3_10_M.1.1      360004.0           12.793873   
A1.MAA000923.3_9_M.1.1       290282.0           12.578611   

                         pct_counts_in_top_50_genes  \
index                                                 
A1.B003290.3_38_F.1.1                     25.884766   
A1.B003728.3_56_F.1.1                     43.051933   
A1.MAA000560.3_10_M.1.1                   44.066006   
A1.MAA000564.3_10_M.1.1                   41.650926   
A1.MAA000923.3_9_M.1.1                    63.392494   

                         pct_counts_in_top_100_genes  \
index                                                  
A1.B003290.3_38_F.1.1                      32.847017   
A1.B003728.3_56_F.1.1                      52.912721   
A1.MAA000560.3_10_M.1.1                    51.514553   
A1.MAA000564.3_10_M.1.1                    48.958623   
A1.MAA000923.3_9_M.1.1                     69.136908   

                         pct_counts_in_top_200_genes  \
index                                                  
A1.B003290.3_38_F.1.1                      42.219573   
A1.B003728.3_56_F.1.1                      65.313309   
A1.MAA000560.3_10_M.1.1                    60.164211   
A1.MAA000564.3_10_M.1.1                    56.997422   
A1.MAA000923.3_9_M.1.1                     77.168753   

                         pct_counts_in_top_500_genes  total_counts_ERCC  \
index                                                                     
A1.B003290.3_38_F.1.1                      59.472666            10201.0   
A1.B003728.3_56_F.1.1                      87.315423            67351.0   
A1.MAA000560.3_10_M.1.1                    74.451190            29870.0   
A1.MAA000564.3_10_M.1.1                    69.736170             2592.0   
A1.MAA000923.3_9_M.1.1                     90.522320            11002.0   

                         log1p_total_counts_ERCC  pct_counts_ERCC  
index                                                              
A1.B003290.3_38_F.1.1                   9.230339         2.615138  
A1.B003728.3_56_F.1.1                  11.117688         8.674345  
A1.MAA000560.3_10_M.1.1                10.304644         1.848292  
A1.MAA000564.3_10_M.1.1                 7.860571         0.719992  
A1.MAA000923.3_9_M.1.1                  9.305923         3.790107  


This is the gene quality control dataframe
                   n_cells_by_counts  mean_counts  log1p_mean_counts  \
index                                                                  
0610005C13Rik                     28     0.118201           0.111721   
0610007C21Rik                   2399   206.211990           5.333742   
0610007L01Rik                    961    35.938549           3.609256   
0610007N19Rik                    272     6.498383           2.014687   
0610007P08Rik                    582    16.328728           2.852366   
...                              ...          ...                ...   
Zzef1                            937    29.685974           3.423806   
Zzz3                             992    27.208467           3.339622   
a                                 10     0.022052           0.021813   
l7Rn6                           1145    48.354897           3.899037   
zsGreen_transgene                  0     0.000000           0.000000   

                   pct_dropout_by_counts  total_counts  log1p_total_counts  
index                                                                       
0610005C13Rik                  99.176713         402.0            5.998937  
0610007C21Rik                  29.461923      701327.0           13.460731  
0610007L01Rik                  71.743605      122227.0           11.713643  
0610007N19Rik                  92.002352       22101.0           10.003424  
0610007P08Rik                  82.887386       55534.0           10.924768  
...                                  ...           ...                 ...  
Zzef1                          72.449280      100962.0           11.522510  
Zzz3                           70.832108       92536.0           11.435364  
a                              99.705969          75.0            4.330733  
l7Rn6                          66.333431      164455.0           12.010398  
zsGreen_transgene             100.000000           0.0            0.000000  

[23433 rows x 6 columns]
</code></pre></div>      </div>
    </div>
  </div>
</div>

<h1 id="quality-control-for-cells">Quality control for cells</h1>

<h2 id="library-size">Library size</h2>
<p>First, we consider the total number of reads detected per cell.
Cells with few reads are likely to have been broken or failed to capture a cell, and should thus be removed.</p>

<div class="cell code_cell">
  <div class="input_area">
    <div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">plt</span><span class="o">.</span><span class="n">hist</span><span class="p">(</span><span class="n">cell_qc_dataframe</span><span class="p">[</span><span class="s">'total_counts'</span><span class="p">],</span> <span class="n">bins</span><span class="o">=</span><span class="mi">1000</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s">'Total counts'</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s">'N cells'</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">axvline</span><span class="p">(</span><span class="mi">50000</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s">'red'</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlim</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mf">1e6</span><span class="p">)</span> <span class="c"># Try plotting with and without scaling the x-axis. When is this helpful?</span>

</code></pre></div>    </div>
  </div>

  <div class="output_wrapper">
    <div class="output_subarea">

      <div class="output_data_text highlighter-rouge"><div class="highlight"><pre class="highlight"><code>(0, 1000000.0)
</code></pre></div>      </div>

    </div>
  </div>
  <div class="output_wrapper">
    <div class="output_subarea">

      <p class="output_png"><img src="../images/preprocessing/01-basic-qc_6_1.png" alt="png" /></p>

    </div>
  </div>
</div>

<p>We have a few cells with fewer than 50,000 reads; we’ll remove these downstream.</p>

<h2 id="detected-genes">Detected genes</h2>

<p>In addition to ensuring sufficient sequencing depth for each sample, we also want to make sure that the reads are distributed across the transcriptome. Thus, we count the total number of unique genes detected in each sample.</p>

<div class="cell code_cell">
  <div class="input_area">
    <div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">plt</span><span class="o">.</span><span class="n">hist</span><span class="p">(</span><span class="n">cell_qc_dataframe</span><span class="p">[</span><span class="s">'n_genes_by_counts'</span><span class="p">],</span> <span class="n">bins</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s">'N genes'</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s">'N cells'</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">axvline</span><span class="p">(</span><span class="mi">1000</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s">'red'</span><span class="p">)</span>

</code></pre></div>    </div>
  </div>

  <div class="output_wrapper">
    <div class="output_subarea">

      <div class="output_data_text highlighter-rouge"><div class="highlight"><pre class="highlight"><code>&lt;matplotlib.lines.Line2D at 0x1086951d0&gt;
</code></pre></div>      </div>

    </div>
  </div>
  <div class="output_wrapper">
    <div class="output_subarea">

      <p class="output_png"><img src="../images/preprocessing/01-basic-qc_9_1.png" alt="png" /></p>

    </div>
  </div>
</div>

<p>From the plot we conclude that most cells have between ~1,000-5,000 detected genes,
which is typical for smartseq2 data. However, this varies by
experimental protocol and sequencing depth.</p>

<p>The most notable feature in the above plot is the little peak on the left hand side of the
distribution. If detection rates were equal across the cells then the
distribution should be approximately normal. Thus, we will remove those
cells in the tail of the distribution (fewer than ~1000 detected genes).</p>

<h2 id="spike-ins">Spike-ins</h2>

<p>Another measure of cell quality is the ratio between ERCC spike-in RNAs and endogenous RNAs. This ratio can be used to estimate the total amount of RNA in the captured cells. Cells with a high level of spike-in RNAs had low starting amounts of RNA, likely due to the cell being dead or stressed which may result in the RNA being degraded.</p>

<h3 id="exercise">Exercise</h3>

<p>Plot the distribution of <code class="highlighter-rouge">pct_counts_ERCC</code> in this dataset. What do you think is a reasonable threshold?</p>

<p>
<details>
<summary><h3>Solution</h3></summary>
<code>plt.hist(cell_qc_dataframe['pct_counts_ERCC'], bins=1000)
plt.xlabel('Percent counts ERCC')
plt.ylabel('N cells')
plt.axvline(10, color='red')</code>
<p>    
Placing a threshold is always a judgement call. Here, the majority of cells have less than 10% ERCC counts, but there's a long tail of cells that have very high spike-in counts; these are likely dead cells and should be removed. 
&lt;/details&gt;



## Cell filtering

Now we can define a cell filter based on our findings:



### Exercise

Use the SCANPY function `sc.pp.filter_cells()` to filter cells with fewer than 50,000 reads. Repeat to remove those with fewer than 750 genes detected. How many cells does this remove?  
_Hint: start with the Parameters list in `help(sc.pp.filter_cells)`

<p>
<details>
<summary><h3>Solution</h3></summary>
<code>print('Started with: \n', adata)
sc.pp.filter_cells(adata, min_genes = 750)
sc.pp.filter_cells(adata, min_counts = 50000)
print('Finished with: \n', adata)</code>
</details>



There isn't an automatic function for removing cells with a high percentage of ERCC reads, but we can use a _mask_ to remove them like so:



<div class="cell code_cell">
          <div class="input_area">
            <div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">low_ERCC_mask</span> <span class="o">=</span> <span class="p">(</span><span class="n">cell_qc_dataframe</span><span class="p">[</span><span class="s">'pct_counts_ERCC'</span><span class="p">]</span> <span class="o">&lt;</span> <span class="mi">10</span><span class="p">)</span>
<span class="n">adata</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[</span><span class="n">low_ERCC_mask</span><span class="p">]</span>

</code></pre></div>            </div>
          </div>

        </div>



## Quality control for genes

It is typically a good idea to remove genes whose expression level is considered __"undetectable"__. We define a gene as  detectable if at least two cells contain more than 5 reads from the gene. However, the threshold strongly depends on the sequencing depth. It is important to keep in mind that genes must be filtered after cell filtering since some genes may only be detected in poor quality cells.



<div class="cell code_cell">
          <div class="input_area">
            <div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">plt</span><span class="o">.</span><span class="n">hist</span><span class="p">(</span><span class="n">gene_qc_dataframe</span><span class="p">[</span><span class="s">'n_cells_by_counts'</span><span class="p">],</span> <span class="n">bins</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s">'N cells expressing &gt; 0'</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s">'N genes'</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">axvline</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s">'red'</span><span class="p">)</span>

</code></pre></div>            </div>
          </div>

          <div class="output_wrapper">
            <div class="output_subarea">

              <div class="output_data_text highlighter-rouge"><div class="highlight"><pre class="highlight"><code>&lt;matplotlib.lines.Line2D at 0x1a29ceccd0&gt;
</code></pre></div>              </div>

            </div>
          </div>
          <div class="output_wrapper">
            <div class="output_subarea">

              <p class="output_png"><img src="../images/preprocessing/01-basic-qc_18_1.png" alt="png" /></p>

            </div>
          </div>
        </div>



<div class="cell code_cell">
          <div class="input_area">
            <div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">plt</span><span class="o">.</span><span class="n">hist</span><span class="p">(</span><span class="n">gene_qc_dataframe</span><span class="p">[</span><span class="s">'total_counts'</span><span class="p">],</span> <span class="n">bins</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s">'Total counts'</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s">'N genes'</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">axvline</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s">'red'</span><span class="p">)</span>
<span class="c"># plt.ylim(0,1000) # try plotting with and without restricting the axis scales. Why is this sometimes helpful?</span>

</code></pre></div>            </div>
          </div>

          <div class="output_wrapper">
            <div class="output_subarea">

              <div class="output_data_text highlighter-rouge"><div class="highlight"><pre class="highlight"><code>&lt;matplotlib.lines.Line2D at 0x1a320a00d0&gt;
</code></pre></div>              </div>

            </div>
          </div>
          <div class="output_wrapper">
            <div class="output_subarea">

              <p class="output_png"><img src="../images/preprocessing/01-basic-qc_19_1.png" alt="png" /></p>

            </div>
          </div>
        </div>



### Exercise

Use the SCANPY function `sc.pp.filter_genes()` to filter genes according to the criteria above. How many genes does this remove?  
_Hint: start with the Parameters list in `help(sc.pp.filter_genes)`

<p>
<details>
<summary><h3>Solution</h3></summary>
<code>print('Started with: \n', adata)
sc.pp.filter_genes(adata, min_cells = 2)
sc.pp.filter_genes(adata, min_counts = 10)
print('Finished with: \n', adata)</code>
</details>



### Save the data



<div class="cell code_cell">
            <div class="input_area">
              <div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c"># print(adata) ## Final dimensions of the QC'd dataset</span>
<span class="n">adata</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">'../data/brain_qc.h5ad'</span><span class="p">)</span>

</code></pre></div>              </div>
            </div>

          </div>



## Big Exercise

Load a demo dataset with `pbmc = sc.read('../data/pbmc3k.h5ad')`, which profiles 3,000 peripheral blood mononuclear cells.
Perform exactly the same QC analysis with this dataset. Compare your results to ours (below).



<div class="cell code_cell">
            <div class="input_area">
              <div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">pbmc</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s">'../data/pbmc3k.h5ad'</span><span class="p">)</span>

</code></pre></div>              </div>
            </div>

          </div>



# Fill me in

</p></p></p></details></p>

              <nav class="c-page__nav">
  
    
    <a id="js-page__nav__prev" class="c-page__nav__prev" href="/scRNA-python-workshop/preprocessing/00-tabula-muris.html">
      〈 <span class="u-margin-right-tiny"></span> Tabula muris
    </a>
  

  
    
    <a id="js-page__nav__next" class="c-page__nav__next" href="/scRNA-python-workshop/preprocessing/02-normalization.html">
      Normalization and PCA <span class="u-margin-right-tiny"></span> 〉
    </a>
  
</nav>

            </div>
          </div>
        </div>
      </main>
    </div>

  </body>
</html>
